pyogrio/_ogr.pyx

```diff
@@ -113,3 +121,130 @@ def ogr_list_drivers():
 
     return drivers
 
+
+cdef void set_proj_search_path(str path):
+    """Set PROJ library data file search path for use in GDAL."""
+    cdef char **paths = NULL
+    cdef const char *path_c = NULL
+    path_b = path.encode("utf-8")
+    path_c = path_b
+    paths = CSLAddString(paths, path_c)
+    OSRSetPROJSearchPaths(<const char *const *>paths)
+
+
+cdef char has_gdal_data():
+    """Verify that GDAL library data files are correctly found.
+
+    Adapted from Fiona (_env.pyx).
+    """
+
+    if CPLFindFile("gdal", "header.dxf") != NULL:
+        return True
+
+    return False
+
+
+cdef char has_proj_data():
+    """Verify that PROJ library data files are loaded by GDAL.
+
+    Returns
+    -------
+    bool
+        True if a test spatial reference object could be created, which verifies
+        that data files are correctly loaded.
+
+    Adapted from Fiona (_env.pyx).
+    """
+    cdef OGRSpatialReferenceH srs = OSRNewSpatialReference(NULL)
+
+    try:
+        exc_wrap_ogrerr(exc_wrap_int(OSRImportFromEPSG(srs, 4326)))
+    except CPLE_BaseError:
+        return 0
+    else:
+        return 1
+    finally:
+        if srs != NULL:
+            OSRRelease(srs)
+
+
+
+def init_gdal_data():
+    """Set GDAL data search directories in the following precedence:
+    - GDAL_DATA env var
+    - wheel copy of gdal_data
+    - default install of GDAL data files
+    - other well-known paths under sys.prefix
+
+    Adapted from Fiona (env.py, _env.pyx).
+    """
+
+    if "GDAL_DATA" in os.environ:
+        set_gdal_config_options({"GDAL_DATA": os.environ["GDAL_DATA"]})
```

Comment:
Not that it gives that much of simplification code-wise, but I am wondering to what extent it is needed to set the option in this case, because GDAL itself already will check for the GDAL_DATA env variable.

Reply:
(but it also doesn't hurt, so my comment doesn't matter that much)

Reply:
This was from Fiona, so I assume there was a good reason for it.

My guess is that it is related to precedence: we want to check this before falling back to checking the wheel location to give greatest control over how it gets set (otherwise, if we left it to GDAL to use env var, it would be after the check for the wheel location).

That said, I'm wondering now if the wheel location should take top precedence; if we deliberately put the data files there, we know they are correct for the version of GDAL pyogrio is compiled against, so it really doesn't make sense to let the user override that.  So then we could fall back to letting GDAL detect this env var.

Reply:
> otherwise, if we left it to GDAL to use env var, it would be after the check for the wheel location).

Yes, we should still return early here if GDAL_DATA is properly set (so my suggestion was only about removing this single line, and not the whole block)

> That said, I'm wondering now if the wheel location should take top precedence

That's a good point, I also can't think of any way that you wouldn't want to use those files _if_ installed from a wheel.

One use case I am thinking about where this could actually matter is the (certainly not recommended! :)) situation of using a conda environment which has GDAL/PROJ installed (eg from installing fiona with conda), and then installing pyogrio from a wheel in that environment. In that case GDAL_DATA will be set and pointing to conda's GDAL, and thus it could give problems if first using the env variable (assuming it is still using the actual libraries included in the wheel).